File: web_1_demo_movies/movieapp/seeded_loader.py
"""
Seeded data loader for web_1_demo_movies (Django).
Loads data from webs_server API using v2-seed for deterministic selection.
"""
import os
import requests
from typing import List, Dict, Any, Optional
from django.conf import settings


def get_api_base_url() -> str:
    """Get the API base URL from environment or settings."""
    api_url = os.environ.get("API_URL") or getattr(settings, "API_URL", None)
    if api_url:
        return api_url.rstrip("/")
    # Default to webs_server in Docker network
    return "http://app:8080"


def is_db_load_mode_enabled() -> bool:
    """Check if V2 DB load mode is enabled."""
    enabled = os.environ.get("ENABLE_DYNAMIC_V2_DB_MODE", "false").lower() == "true"
    if not enabled:
        enabled = getattr(settings, "ENABLE_DYNAMIC_V2_DB_MODE", False)
    return enabled


def fetch_seeded_selection(
    entity_type: str,
    seed_value: int,
    limit: int = 50,
    method: str = "distribute",
    filter_key: Optional[str] = None,
) -> List[Dict[str, Any]]:
    """
    Fetch seeded selection from webs_server API.
    
    Args:
        entity_type: Entity type (e.g., "movies")
        seed_value: Seed value for deterministic selection
        limit: Maximum number of items to return
        method: Selection method (select, shuffle, filter, distribute)
        filter_key: Key to filter on (for filter/distribute methods)
    
    Returns:
        List of dictionaries representing the selected entities
    """
    if not is_db_load_mode_enabled():
        return []
    
    api_url = get_api_base_url()
    project_key = "web_1_demo_movies"
    
    params = {
        "project_key": project_key,
        "entity_type": entity_type,
        "seed_value": seed_value,
        "limit": limit,
        "method": method,
    }
    
    if filter_key:
        params["filter_key"] = filter_key
    
    try:
        response = requests.get(
            f"{api_url}/datasets/load",
            params=params,
            timeout=10,
        )
        
        if response.status_code == 200:
            result = response.json()
            return result.get("data", [])
        else:
            print(f"[seeded_loader] API error: {response.status_code} - {response.text}")
            return []
    except requests.exceptions.RequestException as e:
        print(f"[seeded_loader] Request failed: {e}")
        return []
    except Exception as e:
        print(f"[seeded_loader] Unexpected error: {e}")
        return []


def load_movies_from_api(seed_value: Optional[int] = None, limit: int = 100) -> List[Dict[str, Any]]:
    """
    Load movies from webs_server API using v2-seed.
    
    Args:
        seed_value: Seed value (defaults to 1 if None and v2 is enabled)
        limit: Maximum number of movies to return
    
    Returns:
        List of movie dictionaries
    """
    if not is_db_load_mode_enabled():
        return []
    
    effective_seed = seed_value if seed_value is not None else 1
    
    movies = fetch_seeded_selection(
        entity_type="movies",
        seed_value=effective_seed,
        limit=limit,
        method="distribute",
        filter_key="category",  # Distribute by category if available
    )
    
    return movies

File: web_15_autodiscord/src/shared/seeded-loader.ts
const PROJECT_KEY = "web_15_autodiscord";

function getApiBaseUrl(): string {
  const envUrl = process.env.NEXT_PUBLIC_API_URL || process.env.API_URL;
  const origin = typeof window !== "undefined" ? window.location?.origin : undefined;
  if (envUrl && !envUrl.includes("localhost")) return envUrl;
  if (origin) return `${origin}/api`;
  return envUrl || "http://app:8090";
}

export interface SeededLoadOptions {
  projectKey: string;
  entityType: string;
  seedValue?: number;
  limit?: number;
  method?: "select" | "shuffle" | "filter" | "distribute";
}

function validateOptions(options: SeededLoadOptions): void {
  if (!options.projectKey || typeof options.projectKey !== "string") {
    throw new Error("SeededLoadOptions.projectKey is required");
  }
  if (!options.entityType || typeof options.entityType !== "string") {
    throw new Error("SeededLoadOptions.entityType is required");
  }
}

/**
 * Fetch a seeded slice of entity data from the webs_server API.
 * @throws Error on network failure, non-OK response, or invalid response body
 */
export async function fetchSeededSelection<T = unknown>(options: SeededLoadOptions): Promise<T[]> {
  validateOptions(options);
  const baseUrl = getApiBaseUrl();
  const seed = options.seedValue ?? 1;
  const limit = Math.min(Math.max(1, options.limit ?? 50), 500);
  const method = options.method ?? "select";
  const params = new URLSearchParams({
    project_key: options.projectKey,
    entity_type: options.entityType,
    seed_value: String(seed),
    limit: String(limit),
    method,
  });
  const url = `${baseUrl}/datasets/load?${params.toString()}`;

  let resp: Response;
  try {
    resp = await fetch(url, { method: "GET" });
  } catch (err) {
    const message = err instanceof Error ? err.message : "Network request failed";
    throw new Error(`Seeded selection failed: ${message}`);
  }

  if (!resp.ok) {
    const body = await resp.text().catch(() => "");
    throw new Error(`Seeded selection failed: ${resp.status}${body ? ` — ${body.slice(0, 100)}` : ""}`);
  }

  let json: { data?: unknown };
  try {
    json = await resp.json();
  } catch {
    throw new Error("Seeded selection failed: invalid JSON response");
  }

  const data = json?.data;
  if (!Array.isArray(data)) {
    throw new Error("Seeded selection failed: response data is not an array");
  }
  return data as T[];
}

export async function fetchDiscordEntities<T>(entityType: string, seedValue: number, limit: number): Promise<T[]> {
  return fetchSeededSelection<T>({
    projectKey: PROJECT_KEY,
    entityType,
    seedValue,
    limit,
  });
}
